docs/design/1-overview.md

```diff
@@ -0,0 +1,479 @@
+    And Mark should not receive any error description
+
+  Scenario: Query a job (not_exist)
+    Given Mark has auth certificates in place
+    When Mark requests to query a job
+    And there are no internal system errors
+    And the job does not exist
+    Then Mark should receive an unauthorised response
+    And Mark should not be able to determine if the job existed.
+
+  Scenario: Query a job (stream_output)
+    Given Mark has auth certificates in place
+    When Mark requests to query a job with output stream
+    And there are no internal system errors
+    And the job exists
+    Then Mark should receive a success response
+    And Mark should receive an output stream for the job
+    But the stream should include the entire output history
+  
+```
+
+---
+
+# 🏗️ Architectural Overview
+
+## 🖼️ High-Level Diagram
+
+![1-job-worker-hex-arch.png](images/1-job-worker-hex-arch.png)
+
+## 🛠️ Key Components & Responsibilities
+Describe major modules, services, or layers in the system.
+
+System will be a layered architecture. Domain logic will interact with an abstraction (service) of the lowest level worker package, ensuring the system can be extended.
+
+# Worker Package (Library)
+The worker package will use the following interface structure (subject to refinement);
+
+```go
+const cgroupRoot = "/sys/fs/cgroup" // Adjust based on cgroup version
+
+// Command represents a job request with resource limits.
+type Command struct {
+	Name     string            // Command name (e.g., "ping") 
+	Args     []string          // Command arguments (e.g., ["-c", "4", "google.com"])
+	Env      map[string]string // Environment variable overrides
+	Timeout  time.Duration     // Max execution time before forceful stop
+	CPULimit string            // CPU limit (e.g., "0.5" = 50% of one core)
+	MemLimit string            // Memory limit (e.g., "500M")
+}
+
+// Job represents a running process with cgroup tracking.
+type Job struct {
+	ID     string             // Job identifier
+	Cmd    *exec.Cmd
+	Status *Status
+	cancel context.CancelFunc // Job cancellation
+	Cgroup string             // cgroup path
+	logBuf *LogBuffer         // Captures stdout/stderr for streaming
+}
+
+// Status represents the current state of a Job.
+type Status struct {
+	Pid      int  // Process identifier
+	ExitCode int  // Exit code (-1 if not exited)
+	Exited   bool // Whether the process has exited
+}
+
+// LogBuffer streams logs from a process.
+type LogBuffer struct {
+	mu       sync.Mutex
+	buffer   []string
+	subscribers map[chan string]struct{}
+}
+
+// Worker manages Linux jobs and their lifecycle.
+type Worker interface {
+	Start(command Command) (jobID string, err error)         // Starts a job
+	Stop(jobID string) error                                 // Stops a job
+	Query(jobID string) (status Status, err error)          // Queries a job's status
+	Stream(ctx context.Context, jobID string) (<-chan string, error) // Streams process output
```

Comment:
Could you provide more details on how streaming will be implemented? In particular, this channel-based API looks like it could suffer from slow consumers holding up the producer.

Reply:
I have changed this with an explanation that I'll be using temp files to handle the process output, with a log rotation, and each consumer will read the file(s) independent of each other, solving that problem.

Reply:
How does this change impact the API? Will you still have a channel based API here, or does the move to temp files allow you to present a different API?

What will your notification mechanism be with files? Once a reader catches up to current time, how will they be notified that new data has been written?

Reply:
Moving to temp files allows each client to manage its own reads, and it can operate more like a classic tail command. Each client (through the job service that will sit between the worker and the domain logic) will call `Stream()`. There will no longer be any pub/sub logic encapsulating all clients.

I also don't need to manage an in-memory buffer for replay which is nice, because it means there is less likelihood of missed messages. The file is the source of truth only.

Reply:
Ah yes I didn't submit.

I've changed the channel output to `[]byte`, and will process the file in chunks (likely 4KB). This is what will be streamed to the client. Should fulfill the arbitraty data requirements.

I will continue to use a channel system because it best suits individual clients calling `Stream()`, and has natural backpressure from using an unbuffered channel. I also don't need to manage an in-memory buffer.

Reply:
No problem sticking with the channel so long as you ensure that one slow consumer can't hold up the producer. 

As far as the buffer goes, I don't think waiting for 4KB of data before sending anything to the client is the right UX. Imagine a process that outputs a few bytes per second. It would take 30-60 minutes before the client sees any output. 

Can we do something closer to real time so that the client gets a nice experience no matter how much (or how little) output the job produces?

Reply:
Each client will ultimately call `Stream()` so the blocking won't be a problem. 

Also I agree, I'll update to be more realtime. I was mostly thinking about efficiency because depending on the data type it may cause a lot of requests.